checker/tests/nullness/Issue3150.java

```diff
@@ -1,15 +1,13 @@
-// @skip-test until the bug is fixed
-
 // Test case for https://tinyurl.com/cfissue/3150 .
 
 import org.checkerframework.checker.nullness.qual.NonNull;
 import org.checkerframework.checker.nullness.qual.Nullable;
 
 class Issue3150 {
     void foo(@Nullable Object nble, @NonNull Object nn) {
-        // :: error: (argument.type.incompatible)
+        // :: error: (type.argument.type.incompatible)
```

Comment:
I find this change in expected error message unintuitive and undesirable.
There is no type argument here, so raising an invalid type argument error is misleading.
Type argument inference should always produce type arguments that respect the bounds. Then supplied arguments can cause errors if they don't respect that bound.
Here, I would expect that type argument inference determines `@NonNull Object` as the type argument and then the argument should give an error.

For a similar reason I don't like https://github.com/typetools/jdk/pull/36 : it changes the signature just to account for bad type inference, even though that bad type inference shouldn't give different errors.

Would it be hard to do a GLB between each inferred type argument and the bounds?

Reply:
GLB the inferred type argument and its corresponding bound is tricky when the bound is another type variable. See  [checker/tests/nullness/InferNullType.java](https://github.com/typetools/checker-framework/blob/master/checker/tests/nullness/InferNullType.java).  As you can see in that example, the Checker Framework currently does infer type arguments that are outside its bounds.  I'll spend a little more time trying to change this so that it infers a type argument within the bound, but if it's not easy, then I think we should just open a separate issue.

Reply:
Something like the following should work:

```
1. get possible incorrect assignment of type parameters to inferred type arguments
2. Iterate while there are changes:
2.1 Compute the type parameter bounds under the current assignment, to resolve type parameter uses
2.2 Go through all type parameters and GLB the inferred type argument with the current bound
```

Once #979 is fixed, this additional step shouldn't be necessary, but this might be a good time to improve the quality of type inference.